File: 2017-2018/December/Bronze/TheBovineShuffle.java
/*
ID: srihank1
LANG: JAVA
PROG: template
*/
import java.util.*;
import java.io.*;

public class TheBovineShuffle {

  public static class FastReader {/* Fast class in order to quickly read inputs*/

     BufferedReader br;StringTokenizer st;
      FastReader(InputStream stream) {
         try {br = new BufferedReader(new InputStreamReader(stream));}
         catch (Exception e) {e.printStackTrace();}}
      public FastReader(String str) throws IOException {br = new BufferedReader(new FileReader(str));}
      String next() {
         while (st == null || !st.hasMoreElements()) {
            try {st = new StringTokenizer(br.readLine());}
            catch (IOException  e) {e.printStackTrace();}}
         return st.nextToken();}
      int nextInt() {
         return Integer.parseInt(next());}
      long nextLong() {
         return Long.parseLong(next());}
      double nextDouble() {
         return Double.parseDouble(next());}
      char nextChar() {
         return (next().charAt(0));}
      String nextLine() {String str = "";
         try {str = br.readLine();}
         catch (IOException e) {e.printStackTrace();}
         return str;}}
  public static void println(Object o) {System.out.println(o);}
  public static void print(Object o) {System.out.print(o);}
  public static int Int(String s) {return Integer.parseInt(s);}
  public static int max(int... x) {int curr = x[0];for (int n:x) {curr = Math.max(curr, n);}return curr;}
  public static int min(int... x) {int curr = x[0];for (int n:x) {curr = Math.min(curr, n);}return curr;}


   public static void main(String[] args) throws IOException{
      String input1 = "shuffle.in";
      String input2 = "input.txt";
      PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("shuffle.out")));
      FastReader sc = new FastReader(input1);
      //FastReader sc = new FastReader(System.in);
      int n = sc.nextInt();
      int[] shuffle = new int[n];
      int[] rev = new int[n];
      for (int i = 0; i < n; i++) {
        shuffle[i] = sc.nextInt()-1;
      }
      int[] arr = new int[n];
      for (int i = 0; i < n; i++) {
        arr[i] = sc.nextInt();
      }
      for (int i = 0; i < n; i++) {
        rev[shuffle[i]] = i;
      }
      int[][] floor = new int[3][n];
      for (int i = 0; i < 3; i++) {
        for (int j = 0; j < n; j++) {
          floor[i][rev[j]] = arr[j];
        }
        arr = floor[i].clone();
      }
      for (int iter: arr) {
        pw.println(iter);
      }


      pw.close();

   }
}

File: 2017-2018/December/Bronze/TheBovineShuffle.cpp
#include <bits/stdc++.h>

using namespace std;


int main() {
  // freopen("input.txt", "r", stdin);
  freopen("shuffle.in", "r", stdin);
  freopen("shuffle.out", "w", stdout);
  int n; cin >> n;
  int arr[n];
  for (int i= 0; i < n; i++) {
    int a; cin >> a;a--;
    arr[i] = a;
  }
  int ans[n];
  for (int i =0; i < n; i++) {
    int a = arr[arr[arr[i]]];
    ans[i] = a;
  }
  int sec[n];
  for (int i = 0; i < n; i++) {
    sec[ans[i]] = i;
  }
  int pri[n];
  for (int i = 0; i < n; i++) {
    int a; cin >> a;
    pri[sec[i]] = a;
  }
  for (int i = 0; i < n; i++) {
    cout << pri[i] << endl;
  }

}
/*
0: 0
1: 4
2: 1
3: 2
4: 3
*/
